File: doc-examples/example-java/src/main/java/example/ConsumerProducer.java
package example;

import io.micronaut.pulsar.annotation.PulsarConsumer;
import io.micronaut.pulsar.annotation.PulsarProducer;
import io.micronaut.pulsar.annotation.PulsarSubscription;
import org.apache.pulsar.client.api.SubscriptionType;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

@PulsarSubscription(subscriptionName = "pulsar-jtest-subscription", subscriptionType = SubscriptionType.Shared) // <1>
public class ConsumerProducer { // <2>
    @PulsarConsumer(topic = "persistent://public/default/messages-java-docs", consumerName = "shared-consumer-jtester") // <3>
    public void messagePrinter(String message) { // <4>
        try {
            String changed = report(message).get();
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }
        //...
    }


    @PulsarProducer(topic = "persistent://public/default/reports-java-docs", producerName = "report-producer-java") // <5>
    public CompletableFuture<String> report(String message) { // <6>
        return CompletableFuture.supplyAsync(() -> String.format("Reporting message %s", message)); // <7>
    }
}

File: doc-examples/example-kotlin/src/main/kotlin/example/ConsumerProducer.kt
package example

import io.micronaut.pulsar.annotation.PulsarConsumer
import io.micronaut.pulsar.annotation.PulsarProducer
import io.micronaut.pulsar.annotation.PulsarSubscription
import org.apache.pulsar.client.api.SubscriptionType
import java.util.concurrent.CompletableFuture

@PulsarSubscription(subscriptionName = "pulsar-ktest-subscription", subscriptionType = SubscriptionType.Shared) // <1>
open class ConsumerProducer { // <2>

    @PulsarConsumer(topic = "persistent://public/default/messages-kotlin-docs", consumerName = "shared-consumer-ktester") // <3>
    fun messagePrinter(message: String) { // <4>
        val changed = report(message).get()
        //...
    }


    @PulsarProducer(topic = "persistent://public/default/reports-kotlin-docs", producerName = "report-producer-kotlin") // <5>
    open fun report(message: String): CompletableFuture<String> { // <6>
        return CompletableFuture.supplyAsync { "Reporting message '$message'" } // <7>
    }
}
